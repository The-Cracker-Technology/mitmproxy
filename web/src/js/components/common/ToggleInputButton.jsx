import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import { Key } from '../../utils'

export default class ToggleInputButton extends Component {

    static propTypes = {
        name: PropTypes.string.isRequired,
        txt: PropTypes.string,
        onToggleChanged: PropTypes.func.isRequired,
        checked: PropTypes.bool.isRequired,
        placeholder: PropTypes.string.isRequired,
        inputType: PropTypes.string
    }

    constructor(props) {
        super(props)
        this.state = { txt: props.txt || '' }
    }

    onKeyDown(e) {
        e.stopPropagation()
        if (e.keyCode === Key.ENTER) {
            this.props.onToggleChanged(this.state.txt)
        }
    }

    render() {
        const {checked, onToggleChanged, name, inputType, placeholder} = this.props
        return (
            <div className="input-group toggle-input-btn">
                <span className="input-group-btn"
                      onClick={() => onToggleChanged(this.state.txt)}>
                    <div className={classnames('btn', checked ? 'btn-primary' : 'btn-default')}>
                        <span className={classnames('fa', checked ? 'fa-check-square-o' : 'fa-square-o')}/>
                        &nbsp;
                        {name}
                    </div>
                </span>
                <input
                    className="form-control"
                    placeholder={placeholder}
                    disabled={checked}
                    value={this.state.txt}
                    type={inputType || 'text'}
                    onChange={e => this.setState({ txt: e.target.value })}
                    onKeyDown={e => this.onKeyDown(e)}
                />
            </div>
        )
    }
}
