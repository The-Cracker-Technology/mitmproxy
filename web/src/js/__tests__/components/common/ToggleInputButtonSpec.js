import React from 'react'
import renderer from 'react-test-renderer'
import ToggleInputButton from '../../../components/common/ToggleInputButton'
import { Key } from '../../../utils'

describe('ToggleInputButton Component', () => {
    let mockFunc = jest.fn(),
        toggleInputButton = undefined,
        tree = undefined

    it('should render correctly', () => {
        toggleInputButton = renderer.create(
            <ToggleInputButton checked={true} name="foo" onToggleChanged={mockFunc}
                               placeholder="bar">text</ToggleInputButton>)
        tree = toggleInputButton.toJSON()
        expect(tree).toMatchSnapshot()
    })

    it('should handle keydown and click action', () => {
        toggleInputButton = renderer.create(
            <ToggleInputButton checked={false} name="foo" onToggleChanged={mockFunc}
                               placeholder="bar" txt="txt">text</ToggleInputButton>)
        tree = toggleInputButton.toJSON()
        let mockEvent = {
            keyCode: Key.ENTER,
            stopPropagation: jest.fn()
        }

        tree.children[1].props.onKeyDown(mockEvent)
        expect(mockFunc).toBeCalledWith("txt")

        tree.children[0].props.onClick()
        expect(mockFunc).toBeCalledWith("txt")
    })

    it('should update state onChange', () => {
        // trigger onChange
        tree.children[1].props.onChange({ target: { value: "foo" }})
        // update the tree
        tree = toggleInputButton.toJSON()
        expect(tree.children[1].props.value).toEqual("foo")
    })
})
