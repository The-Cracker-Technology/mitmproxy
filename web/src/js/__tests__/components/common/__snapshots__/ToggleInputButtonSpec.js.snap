// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleInputButton Component should render correctly 1`] = `
<div
  className="input-group toggle-input-btn"
>
  <span
    className="input-group-btn"
    onClick={[Function]}
  >
    <div
      className="btn btn-primary"
    >
      <span
        className="fa fa-check-square-o"
      />
      Â 
      foo
    </div>
  </span>
  <input
    className="form-control"
    disabled={true}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="bar"
    type="text"
    value=""
  />
</div>
`;
